r2/r2/lib/providers/search/solr.py

```diff
@@ -0,0 +1,847 @@
+                for id_ in ids]
+
+
+        batch = etree.Element("delete")
+        batch.extend(deletes)
+        return self.send_documents(batch)
+
+    def batch_lookups(self):                                                                                            
+        try:                                                                                                            
+            self.things = Thing._by_fullname(self.fullnames, data=True,                                                 
+                                             return_dict=False)                                                         
+        except NotFound:                                                                                                
+            if self.use_safe_get:                                                                                       
+                self.things = safe_get(Thing._by_fullname, self.fullnames,                                              
+                                       data=True, return_dict=False)                                                    
+            else:                                                                                                       
+                raise 
+
+    def xml_from_things(self):
+        '''Generate a <batch> XML tree to send to solr for
+        adding/updating/deleting the given things
+        
+        '''
+        add = etree.Element("add")
+        delete = etree.Element("delete")
+        self.batch_lookups()
+        for thing in self.things:
+            try:
+                if thing._spam or thing._deleted:
+                    delete_node = self.delete_xml(thing)
+                    delete.append(delete_node)
+                elif self.should_index(thing):
+                    add_node = self.add_xml(thing)
+                    add.append(add_node)
+            except (AttributeError, KeyError) as e:
+                # Problem! Bail out, which means these items won't get
+                # "consumed" from the queue. If the problem is from DB
+                # lag or a transient issue, then the queue consumer
+                # will succeed eventually. If it's something else,
+                # then manually run a consumer with 'use_safe_get'
+                # on to get past the bad Thing in the queue
+                if not self.use_safe_get:
+                    raise
+                else:
+                    g.log.warning("Ignoring problem on thing %r.\n\n%r",
+                                  thing, e)
+        return (add, delete)
+
+
+    def inject(self, quiet=False):
+        '''Send things to solr. Return value is time elapsed, in seconds,
+        of the communication with the solr endpoint
+        
+        '''
+
+        xml_things = self.xml_from_things()
+
+        cs_time = 0
+        for batch in xml_things:
+
+            if not len(batch):
+                return 0
```

Comment:
This `return` statement breaks out of the `for` loop, which means if the set of things to `add` is empty, the `delete`s won't get uploaded. It should be a `continue` statement instead.